app/assets/javascripts/admin/order_cycles/controllers/order_cycle_exchanges_controller.js.coffee

```diff
@@ -29,13 +28,29 @@ angular.module('admin.orderCycles')
       OrderCycle.removeExchangeFee(exchange, index)
       $scope.order_cycle_form.$dirty = true
 
-    $scope.addSupplier = ($event) ->
-      $event.preventDefault()
-      OrderCycle.addSupplier($scope.new_supplier_id)
-
-    $scope.addDistributor = ($event) ->
-      $event.preventDefault()
-      OrderCycle.addDistributor($scope.new_distributor_id)
-
     $scope.removeDistributionOfVariant = (variant_id) ->
       OrderCycle.removeDistributionOfVariant(variant_id)
+
+    $scope.loadExchangeProducts = (scope, exchange) ->
+      return if scope.enterprises[exchange.enterprise_id].supplied_products_fetched?
+      scope.enterprises[exchange.enterprise_id].supplied_products_fetched = true
+
+      incoming = true if $scope.view == 'incoming'
+      params = { exchange_id: exchange.id, enterprise_id: exchange.enterprise_id, order_cycle_id: $scope.order_cycle.id, incoming: incoming}
+      Product.index params, (products) ->
+        scope.enterprises[exchange.enterprise_id].supplied_products = products
+
+    # Register listeners to capture first toggle open of the products panel of the exchange
+    $scope.exchangeListChanged = ->
+      panelRows = angular.element(".panel-row")
+      for panelRow in panelRows
+        panelCtrl = angular.element(panelRow).controller('panelCtrl')
```

Comment:
This looks a bit funky as well, picking the elements out of the DOM by their CSS selectors, then triggering methods on their controllers...

I'd have to take a bit of a deeper look before I could give a decent suggestion though.

Reply:
yes 🙈 
all I need is to "run some code when user opens the panel"
I could not get to something better than this 🤕

Reply:
ah, I managed to remove this stuff with a ng-init 🎉  
89f83f0f52802b6e7e7b1d29394f805ef589f688

Reply:
Will the ng-init in that last commit load all the panel data when the page loads instead of on click?

Reply:
no, because the panel template is only loaded when the panel is opened. I tested this :+1: